### PR TITLE
Add a test before download Taxonomies

### PR DESCRIPTION
**Pull request by dobriseb**
## Description

Compare Last modified date of the fils on the server ans the Last download date of the file.
Only upload the file if last modified date > last download date and last modified date > 0

 ## Checklist
 
Make sure you've done all the following (You can delete the checklist before submitting)
 
 - [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.
 - [x] Code is well documented
 - [x] Include unit tests for new functionality
 - [x] All user-visible strings are made translatable
 - [ ] Code passes Travis builds in your branch
 - [x] If you have multiple commits please combine them into one commit by squashing them.
 - [x] Read and understood the contribution guidelines .

### CODE REVIEW

**Comment by teolemon** [comment]
Can you rebase on latest master ?

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/repositories/ProductRepository.java [hunk 1]
```diff
@@ -244,26 +264,81 @@ private ProductRepository() {
...
+    }
+
+    /**
+     * This function update the lastDLtaxonomy setting
+     * @param taxonomy  Name of the taxonomy (allergens, additives, categories, countries, ingredients, labels, tags)
+     * @param lastDL    Date of last update on Long format
+     */
+    public void updateLastDL(String taxonomy, Long lastDL){
+        SharedPreferences mSettings = OFFApplication.getInstance().getSharedPreferences("prefs", 0);
+        mSettings.edit().putLong("lastDL" + taxonomy, lastDL).apply();
```
Comment (teolemon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/repositories/ProductRepository.java]: can we avoid abbreviating things ? lastDL >>lastDownload for instance ?

**Comment by deniger** [comment]
Some magic numbers should be avoid ( -1 for internet issue, 0 not active,...) and should be replaced by explicit constant.
Otherwise, the code is correct for me.
Maybe we could merge it and then modify these feedbacks

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/repositories/ProductRepository.java [hunk 2]
```diff
@@ -134,9 +139,12 @@ private ProductRepository() {
      */
     @Override
     public Single<List<Label>> getLabels(Boolean refresh) {
-        if (refresh || tableIsEmpty(labelDao)) {
+        Long lastModifiedDate = UpdateSinceLastUpload("labels");
+        if (lastModifiedDate > 1) {
```
Comment (deniger) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/repositories/ProductRepository.java]: Ideal: replace Long by a simple class containing an Enum ( ERROR, UPTODATE, TO_DOWNLOAD and SKIPPED) and the timestamp.
It's more explicit

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/views/splash/SplashPresenter.java [hunk 3]
```diff
@@ -36,6 +36,20 @@ public void refreshData() {
                 settings.edit()
                         .putBoolean("firstRun", false)
                         .apply();
+                //Set to 1 the lastDownload for the taxonomy to be loaded :
+                //This two are needed for all flavors :
+                settings.edit().putLong("lastDownloadcategories", Long.valueOf(1)).apply();
```
Comment (deniger) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/views/splash/SplashPresenter.java]: Issue here: if first run is false ( it's the case for all old users), the settings are not intialized

`if(!settings.contains("lastDownloadcategories")){
             if(!settings.contains("lastDownloadcategories")){
            settings.edit().putLong("lastDownloadcategories", Long.valueOf(ProductRepository.TAXONOMY_UP_TO_DATE)).apply();
            settings.edit().putLong("lastDownloadtags", Long.valueOf(ProductRepository.TAXONOMY_UP_TO_DATE)).apply();
            if (BuildConfig.FLAVOR.equals("off") || BuildConfig.FLAVOR.equals("obf")) {
                settings.edit().putLong("lastDownloadadditives", Long.valueOf(ProductRepository.TAXONOMY_UP_TO_DATE)).apply();
                settings.edit().putLong("lastDownloadcountries", Long.valueOf(ProductRepository.TAXONOMY_UP_TO_DATE)).apply();
                settings.edit().putLong("lastDownloadlabels", Long.valueOf(ProductRepository.TAXONOMY_UP_TO_DATE)).apply();
                //For the moment, ingredients are only need if Diet is use.
                settings.edit().putLong("lastDownloadingredients", Long.valueOf(ProductRepository.TAXONOMY_NOT_TO_BE_LOADED)).apply();
            }
            if (BuildConfig.FLAVOR.equals("off")) {
                settings.edit().putLong("lastDownloadallergens", Long.valueOf(ProductRepository.TAXONOMY_UP_TO_DATE)).apply();
            }
        }`

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/repositories/ProductRepository.java [hunk 4]
```diff
@@ -244,26 +271,81 @@ private ProductRepository() {
...
+     * @param taxonomy The name of the taxonomy to be tested
+     *        (allergens, additives, categories, countries, ingredients, labels, tags)
+     *
+     * @return
+     *     TAXONOMY_NO_INTERNET (-1)        no internet connexion.
+     *     TAXONOMY_NOT_TO_BE_LOADED (0)    taxonomy is not marked to be load.
+     *     TAXONOMY_UP_TO_DATE (1)          taxonomy is up to date.
+     *     other :                          date of the new taxonomy on the servers => to be updated
+     */
+    public Long UpdateSinceLastUpload(String taxonomy) {
```
Comment (deniger) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/repositories/ProductRepository.java]:
```suggestion
    public Long updateSinceLastUpload(String taxonomy) {
```

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/repositories/ProductRepository.java [hunk 5]
```diff
@@ -244,26 +271,81 @@ private ProductRepository() {
...
+        Log.i("INFO_URL", "UpdateSinceLastUpload for : " + taxonomy + " begin.");
+        SharedPreferences mSettings = OFFApplication.getInstance().getSharedPreferences("prefs", 0);
+        Long lastDownload = mSettings.getLong("lastDownload" + taxonomy, 0);
+        if (lastDownload > 0) {
```
Comment (deniger) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/repositories/ProductRepository.java]:
```suggestion
        if (lastDownload >TAXONOMY_NOT_TO_BE_LOADED) {
```

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/repositories/ProductRepository.java [hunk 6]
```diff
@@ -244,26 +271,81 @@ private ProductRepository() {
...
+        return TAXONOMY_NOT_TO_BE_LOADED;
+    }
+
+    /**
+     * This function update the lastDownloadtaxonomy setting
+     * @param taxonomy  Name of the taxonomy (allergens, additives, categories, countries, ingredients, labels, tags)
+     * @param lastDownload    Date of last update on Long format
+     */
+    public void updateLastDownload(String taxonomy, Long lastDownload){
+        SharedPreferences mSettings = OFFApplication.getInstance().getSharedPreferences("prefs", 0);
```
Comment (deniger) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/repositories/ProductRepository.java]:
```suggestion
        SharedPreferences mSettings = OFFApplication.getInstance().getSharedPreferences("prefs", TAXONOMY_NOT_TO_BE_LOADED);
```

**Comment by deniger** [comment]
image

**Pull request merged**